### PR TITLE
Add Caliper and Adiak support

### PR DESCRIPTION
**Pull request by daboehme**
Add support for profiling with Caliper. Also includes Adiak support, which adds program metadata (input config etc.) to Caliper profiles. Build with 

```
cmake -DENABLE_CALIPER=On \ 
  -Dcaliper_DIR=<caliper-install-dir>/share/cmake/caliper \
  -DENABLE_ADIAK=On \
  -Dadiak_DIR=<adiak-install-dir>/lib/cmake/adiak
```

Use the `-caliper_config` option in comb to run with Caliper profiling, e.g. `-caliper_config runtime-report`.

### CODE REVIEW

File: src/comb.cpp [hunk 1]
```diff
@@ -547,7 +567,16 @@ int main(int argc, char** argv)
...
+      } else if (strcmp(&argv[i][1], "caliper_config") == 0) {
+      if (i+1 < argc && argv[i+1][0] != '-') {
+        caliper_config = argv[++i];
+#ifndef COMB_ENABLE_CALIPER
+        fgprintf(FileGroup::err_master, "Caliper is not enabled, ignoring caliper_config.\n");
+#endif
       } else {
+        fgprintf(FileGroup::err_master, "No argument to option, ignoring %s.\n", argv[i]);
+      }
+    } else {
```
Comment (MrBurmark) [inline src/comb.cpp]: indentation

**Comment by MrBurmark** [comment]
What data should be added to adiak? The basic configuration is in there, but what about things like combinations of arguments run and their timings?

**Comment by daboehme** [comment]
Hi @MrBurmark , as for what to add to adiak, broadly speaking "any kind of environment/configuration info that might affect performance". That certainly includes the config settings. Other things may be build info like the MPI and CUDA version or the compiler (though compiler might be less relevant for comb). We also often record global performance values like total runtime. We can add these incrementally as needed.

What do you mean by "combinations of arguments run"? Can you have multiple different configurations in the same run?

**Comment by MrBurmark** [comment]
Right, multiple different communication policies, execution policies, and memory spaces can be enabled and run in a single execution of the program.

**Comment by daboehme** [comment]
OK. Adiak should only record information that is constant for the entire program run. If there is a need to distinguish between the different execution policies etc. for performance analysis it's better to use Caliper regions / Comb ranges for that. Adiak could still capture the set of enabled things of course.

File: README.md [hunk 2]
```diff
@@ -51,6 +51,8 @@ The cmake configuration options change which execution patterns and memory space
   - __ENABLE_OPENMP__ Allow use of openmp and enable test combinations using openmp
   - __ENABLE_CUDA__  Allow use of cuda and enable test combinations using cuda
   - __ENABLE_RAJA__  Allow use of RAJA performance portability library
+  - __ENABLE_CALIPER__ Allow use of the Caliper performance profiling library
+  - __ENABLE_ADIAK__ Allow use of the Adiak library for recording program metadata
```
Comment (MrBurmark) [inline README.md:55]: Are there any version requirements to document?

**Comment by MrBurmark** [comment]
Which policies are enabled/disabled is constant for the whole run and affects what is run later, but the actual data would be captured in Caliper. I just wanted to be sure that we're not missing anything.

File: README.md (hunk 2 above)
Comment (daboehme) [reply README.md:55]: Should work with any Adiak version and any recent Caliper version (v2.2+, which was released in 2019).

**Comment by daboehme** [comment]
> Which policies are enabled/disabled is constant for the whole run and affects what is run later, but the actual data would be captured in Caliper. I just wanted to be sure that we're not missing anything.

Indeed, looks like we capture the execution policies that are run through the range annotations. Caliper gets them from there. We should be fine then.

**Pull request merged**